### DISCUSSION TITLE
Fields with an "on create - set default value" action that can be at the same time marked as required

### DISCUSSION THREAD

**Discussion by josias-r** [directus/directus#14134] [Feature Requests]
Hi guys

I'm curious about your opinion on this. Currently you can't really combine "required" fields with a "on create" action. The reason being that when you create a new item (say with a field `owner_user` and an `on create` action to set that fields value to the current user), during the creation phase the fields will still be empty and therefore the "required" rule will kick in and block the user from creating it.

For the use-cases I had so far this is not the preferred way of handling it. I would expect the field to show a message if no value is selected that goes like "If you don't select a user, your current user will be set"

If there are other use-cases where this could make sense (I don't know any), there could also be a new flag for fields like a `soft required` next to the default `required` that will allow an empty value if it will get generated during generation

**Comment by rijkvanzanten** [comment]
I agree. I think it's time for a little bigger revision on how required works exactly, especially on how it interacts with not-nullable fields 🙂

**Reply by josias-r** [reply]
Yes that could make sense 😄
I think the two views that need to be considered is the "dev" and the "ui/user".

As a dev I prefer actual not-nullable fields (for type safety), while through the UI I don't wan't to be blocked by that limitation.

The possible patterns I see are the following:
- The Front-end fetches the data and fills it in
   - Pros:
      -  little complexity
      - Easy to handle in the FE to display the value that will be added by default
      - Keeps the field truly not-nullable
      - Can handle errors/validation **before** data is created
   - Cons:
      - Neglects the API, what if the same "on-create" should happen through the API?
      - Seems more fragile because bugs in the FE can now also affect the values that end up in the DB
- Some sort of middle-ware that run during "on-create" and fills in the fields before it get's saved to the DB **server-side**
   -  Pros:
      - Works for both API and FE
      - Can handle errors/validation very well
      - Keeps the field truly not-nullable
   - Cons:
      - Probably more complex to implement
      - FE doesn't know what the default value will be, and can only display a generic message like "This field will default to your user"
         - Unless the FE also fetches the default data as well, with the downside of added complexity
- On-create actually runs immediately **after** creation and hydrates default values
   - Pros:
      - Easier to implement that a middleware
      - Works for the FE and BE
   - Cons:
      - Seems less flexible
      - Bugs/error handling would be much more critical and hard, because the data was already added to the DB, therefore some sort of "cleanup" would need to happen
      - Fields wouldn't truly be not-nullable -> neglects the dev. The dev really needs to trust directus 100% that everything wen well
      - I really think this is a bad idea 😋

**Reply by rijkvanzanten** [reply]
Here's some immediate thoughts on the possibilities:

> The Front-end fetches the data and fills it in

I think that's a no-go, for the first con you mentioned: It would exclude the API completely. This sort of logic should definitely be handled API-side, which brings us to:

> Some sort of middle-ware that run during "on-create" and fills in the fields before it get's saved to the DB server-side

That's technically how it works now! The on-create stuff is triggered before it's saved to the database. That being said, the `required` validation step runs _before_ that data is inserted, causing the confusing that we have now.

> On-create actually runs immediately after creation and hydrates default values

I second the _I really think this is a bad idea_ statement based on the mentioned cons 🙂

**Reply by josias-r** [reply]
> I think that's a no-go, for the first con you mentioned: It would exclude the API completely. This sort of logic should definitely be handled API-side

Totally agree with that, since Directus is very API focus, this only makes sense.

> The on-create stuff is triggered before it's saved to the database. That being said, the required validation step runs before that data is inserted

If there was an additional "prepare data" step (which get's the default data) _before_ validation happens (and finally stored to the DB) then the creation flow would be very intuitive and robust.
Without a detailed insight into your current system, this seems like a great and feasible approach

> I second the I really think this is a bad idea statement based on the mentioned cons 🙂

Haha. My main concern with such an approach is the generating of invalid data. And once this data is generated it adds a lot of complexity to clean it up. If the fields are already not-nullable on an SQL layer, even manual DB changes (or imports) should prevent invalid states to occur (in general just the mindset of preventing invalid data to ever be possible to be generated, so you don't have to deal with the cases when it happened).
And this added complexity also means more points of failure -> generally making the the system more unstable.
But there might be very smart technical solutions for this that I'm not aware of 😁

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
